src/libxtp/tools/mol2orb.cc

```diff
@@ -0,0 +1,272 @@
+  _aux_basisset_name = options.get(".auxbasisset").as<std::string>();
+}
+
+void Mol2Orb::setupBasis(Orbitals& orbitals) {
+  // We only need the basis set for checking correctness
+  BasisSet bs;
+  bs.Load(_basisset_name);
+
+  // fill DFT AO basis by going through all atoms
+  _basis.Fill(bs, orbitals.QMAtoms());
+  orbitals.setDFTbasisName(_basisset_name);
+  XTP_LOG(Log::error, _log)
+      << TimeStamp() << " Basisset size: " << _basis.AOBasisSize()
+      << std::flush;
+  orbitals.setBasisSetSize(_basis.AOBasisSize());
+  orbitals.setAuxbasisName(_aux_basisset_name);
+}
+
+inline std::string Mol2Orb::readAtoms(QMMolecule& mol, std::string units,
+                                      std::ifstream& input_file) const {
+  std::string line;
+  std::istringstream iss(" ");
+  while (std::getline(input_file, line)) {
+    boost::trim(line);
+    if (line == "" || line[0] == '[') {
+      return line;
+    }
+    iss.str(line);
+    iss.clear();
+
+    // extract data
+    double x, y, z;
+    int atom_id;
+    std::string junk;
+    std::string atom_type;
+    iss >> atom_type >> atom_id >> junk >> x >> y >> z;
+    atom_id =
+        atom_id - 1;  // molden uses indexing from 1, we use indexing from 0
+
+    // Add data to orbitals object
+    using Atom =
+        typename std::iterator_traits<QMMolecule::iterator>::value_type;
+    Eigen::Vector3d pos(x, y, z);
+    if (units == "Angs") {
+      pos = tools::conv::ang2bohr * pos;
+    }
+    mol.push_back(Atom(atom_id, atom_type, pos));
+  }
+  return "";
+}
+
+inline std::vector<std::array<int, 2>> Mol2Orb::getTwoCycles(Index numFunc) {
+  switch (numFunc) {
+    case 1:
+      return _twoCyclesS;
+    case 3:
+      return _twoCyclesP;
+    case 5:
+      return _twoCyclesD;
+    case 7:
+      return _twoCyclesF;
+    case 9:
+      return _twoCyclesG;
+    default:
+      throw std::runtime_error("Impossible number of functions in shell");
+  }
+}
+
+void Mol2Orb::reorderOrbitals(Eigen::MatrixXd& v) {
+  std::vector<Index> reorder;
+  std::vector<Index> multiplier;
+  reorder.reserve(_basis.AOBasisSize());
+  multiplier.reserve(_basis.AOBasisSize());
+
+  // reorder and get multiplier vector
+  Index currentFunction = 0;
+  Index nrOfFunctions = 0;
+  for (const AOShell& shell : _basis) {
+    // Make multiplier vector
+    std::vector<Index> shellmultiplier{
+        _multipliers.begin() + shell.getOffset(),
+        _multipliers.begin() + shell.getOffset() + shell.getNumFunc()};
+    multiplier.insert(multiplier.end(), shellmultiplier.begin(),
+                      shellmultiplier.end());
+    // reorder
+    nrOfFunctions = shell.getNumFunc();
+    for (auto& cycle : getTwoCycles(nrOfFunctions)) {
+      v.row(currentFunction + cycle[0]).swap(v.row(currentFunction + cycle[1]));
+    }
+    currentFunction += nrOfFunctions;
+  }
+
+  // Multiply by multiplier
+  for (Index i = 0; i < v.cols(); i++) {
+    v.row(i) = multiplier[i] * v.row(i);
+  }
+}
```

Comment:
okay this is basically the same stuff that I merged into qmpackage. Maybe we should extract it into a separate reorder class?